tests/tracer/telemetry/test_writer.py

```diff
@@ -0,0 +1,226 @@
+    # Ensure events queue is empty
+    events = telemetry_writer._events_queue
+    assert len(events) == 0
+
+    # Add 1 event to the queue
+    telemetry_writer.app_started_event()
+    # Queue two integrations
+    telemetry_writer.add_integration("integration-1", True)
+    telemetry_writer.add_integration("integration-2", False)
+    # Ensure two integrations have been queued
+    assert len(telemetry_writer._integrations_queue) == 2
+    # Ensure events queue has 2 events (app started and app closed)
+    assert len(telemetry_writer._events_queue) == 1
+    with mock.patch("ddtrace.internal.telemetry.writer.log") as log:
+        # Send all events to the agent proxy
+        telemetry_writer.periodic()
+        # Assert no error logs were generated while sending telemetry requests
+        log.warning.assert_not_called()
+    # Ensure all events have been sent
+    assert len(telemetry_writer._events_queue) == 0
+    assert len(telemetry_writer._integrations_queue) == 0
+
+    # Queue 2 more integrations
+    telemetry_writer.add_integration("integration-3", True)
+    telemetry_writer.add_integration("integration-4", False)
+    assert len(telemetry_writer._integrations_queue) == 2
+    with mock.patch("ddtrace.internal.telemetry.writer.log") as log:
+        # Send both integrations to the agent proxy
+        telemetry_writer.periodic()
+        # Assert no error logs were generated while sending telemetry requests
+        log.warning.assert_not_called()
+    # Ensure all integrations were sent
+    assert len(telemetry_writer._integrations_queue) == 0
+
+
+@httpretty.activate()
+def test_send_request_400(telemetry_writer):
+    """asserts that an warning is logged when a 400 response is returned by the http client"""
+    httpretty.register_uri(httpretty.POST, TEST_URL, status=400)
+    with mock.patch("ddtrace.internal.telemetry.writer.log") as log:
+        # Sends failing app-closed event
+        telemetry_writer.shutdown()
+        # Assert 400 status code was logged
+        log.warning.assert_called_with("failed to send telemetry to the Datadog Agent. response: %s", 400)
+
+
+@httpretty.activate()
+def test_send_request_500(telemetry_writer):
+    """asserts that an warning is logged when a 400 response is returned by the http client"""
+    httpretty.register_uri(httpretty.POST, TEST_URL, status=500)
+    with mock.patch("ddtrace.internal.telemetry.writer.log") as log:
+        # Sends failing app-closed event
+        telemetry_writer.shutdown()
+        # Assert 500 status code was logged
+        log.warning.assert_called_with("failed to send telemetry to the Datadog Agent. response: %s", 500)
+
+
+def test_send_request_exception(telemetry_writer):
+    """asserts that an error is logged when an exception is raised by the http client"""
+    # Create a telemetry writer with an invalid agent url.
+    # This will raise an HttpConnectionException on send_request
+    telemetry_writer = TelemetryWriter("")
```

Comment:
I don't think this is failing how you'd expect it to.

You mentioned adding a test that explicitly raises an exception when parsing the response body of the mocked http call. Does that not work here?

Reply:
Oh this is passing because the agent isn't running on localhost:8126 or the agent doesn't have the agent proxy. I'll fix it by setting agent_url to a host that doesn't exist